File: list/LinkedListReversal.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class LinkedListReversal {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* reversedHead = NULL;
        ListNode* current = head;
        while (current != NULL) {
            ListNode* next = current->next;
            current->next = reversedHead;
            reversedHead = current;
            current = next;
        }
        return reversedHead;
    }

    ListNode* reverseList(ListNode* head) {
        if (head == NULL || head->next == NULL) {
            return head;
        }
        ListNode* reversed = reverseList(head->next);
        head->next->next = head;
        head->next = NULL;
        return reversed;
    }
};

File: list/LinkedListReversal.java

public class LinkedListReversal {
    // Iterative solution; Time complexity O(n), space O(1).
    public ListNode reverse(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode current = head;
        ListNode revHead = null;
        while (current != null) {
            ListNode next = current.next;
            current.next = revHead;
            revHead = current;
            current = next;
        }
        return revHead;
    }

    // Recursive solution; Time complexity O(n), space O(n).
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode current = reverseList(head.next);
        // reverse the current node
        head.next.next = head;
        // dereference the front pointer to the current head
        head.next = null;
        // return the front
        return current;
    }
}
